node/node-sdk.md

```diff
@@ -0,0 +1,187 @@
+# NodeJS SDK Proposal
+
+NodeJS is a perfect candidate for authoring Temporal workflows, by leveraging V8 isolates for running workflow code,
+we can create a JS runtime that's completely deterministic.
+
+The SDK steers developers to write their workflows and activities in TypeScript but vanilla JS is also supported.
+All examples in this proposal are written in TypeScript.
+
+## Table Of Contents
+- [Activities](#activities)
+    - [Examples](#examples)
+    - [Registration](#registration)
+    - [Payloads](#payloads)
+- [Workflows](#workflows)
+    - [Example](#example)
+    - [Activity Options](#activity-options)
+    - [Activity Type Declarations](#activity-type-declarations)
+    - [Registration](#registration-1)
+    - [Limitations](#limitations)
+    - [Signals](#signals)
+    - [Queries](#queries)
+    - [Logging](#logging)
+    - [Sessions](#sessions)
+    - [Deterministic Time](#deterministic-time)
+    - [Error Handling](#error-handling)
+    - [Cancellation](#cancellation)
+    - [Long Histories](#long-histories)
+
+
+## Activities
+Activity implementations are just JS functions, they run on the default NodeJS runtime, along with the Temporal SDK, and can import arbitrary npm packages.
+Activity functions may optionally accept an `ActivityContext` as the first parameter.
+
+### Examples
+#### Synchronous function
+```typescript
+import crypto from 'crypto';
+
+export function sha256(txt: string) {
+  return crypto.createHash('sha256').update(txt).digest('hex');
+}
+```
+
+#### Asynchronous function
+```typescript
+import axios from 'axios';
+
+export async function httpGet(url: string) {
+  const response = await axios.get(url);
+  return response.data;
+}
+```
+
+#### With `ActivityContext`
+```typescript
+import { ActivityContext } from 'temporal-sdk';
+import axios from 'axios';
+
+export async function httpGet(ctx: ActivityContext, url: string) {
+  const response = await axios.get(url, {
+    onDownloadProgress: (progressEvent) => {
+      const total = parseFloat(progressEvent.currentTarget.responseHeaders['Content-Length']);
+      const current = progressEvent.currentTarget.response.length;
+      const progress = Math.floor(current / total * 100);
+      ctx.heartbeat(progress);
+    },
+  });
+  return response.data;
+}
+```
+
+### Registration
+Activities should be registered in their corresponding workers.
+
+```typescript
+worker.registerActivity('httpGet', httpGet, { passContext: true });
+```
+
+### Payloads
```

Comment:
We should support the pluggable DataConverter from the beginning. See [Java](https://github.com/temporalio/sdk-java/tree/master/temporal-sdk/src/main/java/io/temporal/common/converter) and [Go](https://github.com/temporalio/sdk-go/tree/master/converter) implementations of it.